rule/time_date.go

```diff
@@ -77,15 +92,179 @@ func (w lintTimeDate) Visit(n ast.Node) ast.Visitor {
+						fieldName, providedValue,
+					),
+				})
+			}
+
+			if fieldName == "year" {
+				// store the year value for further checks with day
+				parsedDate.year = parsedValue
+
+				// no checks for year, as it can be any value
+				// a year can be negative, zero, or positive
+				continue
+			}
+
+			if hasNegativeSign {
+				w.onFailure(lint.Failure{
+					Category:   "time",
+					Node:       arg,
+					Confidence: 0.5,
+					Failure: fmt.Sprintf(
+						"time.Date %s argument is negative: %s",
+						fieldName, providedValue,
+					),
+				})
+				continue
+			}
+
+			boundaries, ok := timeDateArgumentBoundaries[fieldName]
+			if !ok {
+				// no boundaries for this field, skip it
+				continue
+			}
+			minValue, maxValue := boundaries[0], boundaries[1]
+
+			switch fieldName {
+			case "month":
+				parsedDate.month = parsedValue
+
+				if parsedValue == 0 {
+					// this is a special case, where the month is 0
+					// Go will ignore it and use January
+					// but we can still report it as a failure
+
+					w.onFailure(lint.Failure{
+						Category:   "time",
+						Node:       arg,
+						Confidence: 0.8,
+						Failure:    "time.Date month argument should not be zero",
+					})
+					continue
+				}
+
+			case "day":
+
+				switch {
+				case parsedValue == 0:
+					// this is a special case, where the day is 0
+					// Go will ignore it and use the first day of the month
+					// but we can still report it as a failure
+
+					w.onFailure(lint.Failure{
+						Category:   "time",
+						Node:       arg,
+						Confidence: 0.8,
+						Failure:    "time.Date day argument should not be zero",
+					})
+					continue
+
+				// the month is valid, we can check the day
+				case parsedDate.month >= 1 && parsedDate.month <= 12:
+					// the month is valid, we can check the day
+
+					month := time.Month(parsedDate.month)
+
+					maxValue = w.daysInMonth(parsedDate.year, month)
+
+					var periodInfo string
+					if month == time.February {
+						periodInfo += " " + strconv.FormatInt(parsedDate.year, 10)
+					}
+
+					if parsedValue > maxValue {
+						// we can provide a more detailed error message
+						w.onFailure(lint.Failure{
+							Category:   "time",
+							Node:       arg,
+							Confidence: 0.8,
+							Failure: fmt.Sprintf(
+								"time.Date day argument is %d, but %s has only %d days",
+								parsedValue, month.String()+periodInfo, maxValue,
+							),
+						})
+						continue
+					}
+
+				// possible day and month swap
+				case parsedDate.month > 12 && parsedDate.month <= 31 && parsedValue <= 12:
+					// this is a special case, where the month and day arguments are swapped
+					// for example: time.Date(2023, 31, 6, 0, 0, 0, 0, time.UTC)
+
+					realMonth, realDay := parsedValue, parsedDate.month
+
+					// we must ensure that the date is valid
+					maxValue = w.daysInMonth(parsedDate.year, time.Month(realMonth))
+
+					if realDay >= 1 && realDay <= maxValue {
```

Comment:
self review without a laptop

FIXME: define a new variable and use it

Otherwise, the check with max value can be invalid in the following if statement could be invalid.

TODO: add a test about it

2023-29-2
2023-31-6

I feel like it falls through the net otherwise